# Titanic Data Set - Statistics Review

[Variable Descriptions](https://www.kaggle.com/c/titanic/data).

In [1]:
# data analysis and wrangling
import pandas as pd
import numpy as np
import random as rnd

# visualization
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
df = pd.read_csv('titanic.csv')

**Describe the data. **
- How big?
- What are the columns and what do they mean?

In [3]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [4]:
df.shape

(891, 12)

In [5]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

**What’s the average age of:**

- Any Titanic passenger
- A survivor
- A non-surviving first-class passenger
- Male survivors older than 30 from anywhere but Queenstown

In [64]:
df.groupby('Survived')['Age'].mean()

Survived
0    30.626179
1    28.343690
Name: Age, dtype: float64

In [63]:
df.groupby(['Survived', 'Pclass'])['Age'].mean()

Survived  Pclass
0         1         43.695312
          2         33.544444
          3         26.555556
1         1         35.368197
          2         25.901566
          3         20.646118
Name: Age, dtype: float64

In [65]:
conditional_mask = (df['Age'] > 30) & (df['Embarked'] != 'Q')
df.loc[conditional_mask].groupby(['Survived', 'Sex'])['Age'].mean()

Survived  Sex   
0         female    42.470588
          male      43.771242
1         female    42.176829
          male      41.487805
Name: Age, dtype: float64

In [66]:
# Printout of final answers
print('Average age of:\n')
print('Any Titanic passenger: %.2f' % df['Age'].mean())
print('A survivor: %.2f' % df.groupby('Survived')['Age'].mean()[1])
print('A non-surviving first-class passenger: %.2f' % df.groupby(['Survived', 'Pclass'])['Age'].mean()[0][1])
print('Male survivors older than 30 from anywhere but Queenstown: %.2f' %\
      df.loc[conditional_mask].groupby(['Survived', 'Sex'])['Age'].mean()[1][1])


Average age of:

Any Titanic passenger: 29.70
A survivor: 28.34
A non-surviving first-class passenger: 43.70
Male survivors older than 30 from anywhere but Queenstown: 41.49


**For the groups from the previous task, how far (in years) are the average ages from the median ages?**

In [ ]:
Tit_med = df['Age'].median()
print('Median age of all passengers: %f' % Tit_med)
print('-------------------------')
print('Years from median:')
print('Any Titanic passenger: %f' % (df['Age'].mean() - Tit_med))
print('A survivor: %f' % (survivors['Age'].mean() - Tit_med))
print('A non-surviving first-class passenger: %f' % (ns_firstclass['Age'].mean() - Tit_med))
print('Male survivors older than 30 from anywhere but Queenstown: %f' %\
      (male_surv_30plus_notQ['Age'].mean() - Tit_med))

**What’s the most common:**
- Passenger class
- Port of Embarkation
- Number of siblings or spouses aboard for survivors

In [ ]:
df.describe(include=['O'])

In [ ]:
df['Pclass'].value_counts()

In [ ]:
survivors['SibSp'].value_counts()

**Within what range of standard deviations from the mean (0-1, 1-2, 2-3) is the median ticket price? Is it above or below the mean?**

In [ ]:
df['Fare'].describe()

In [ ]:
df['Fare'].mean() - (df['Fare'].std()/2)

In [ ]:
(df['Fare'].mean() - (df['Fare'].std()/2)) + df['Fare'].std()

A: 0-1 SD range is from 7.35-57.05 and 14.45 (median) lies in this range.

**How much more expensive was the 90th percentile ticket than the 5th percentile ticket? Are they the same class?**

In [ ]:
df['Fare'].quantile(0.9) - df['Fare'].quantile(0.05)

In [ ]:
df[df['Fare'] == df['Fare'].quantile(0.9)]

In [ ]:
df[df['Fare'] == df['Fare'].quantile(0.05)]

A: 90th percentile ticket costs 70.73 more than the 5th percentile fare. More expensive fare passengers are tagged as "Upper" class, cheaper fare passengers as "Lower" class.

**The highest average ticket price was paid by passengers from which port? Null ports don’t count.**

In [ ]:
df[df['Embarked'] == 'C'].mean()

In [ ]:
df[df['Embarked'] == 'Q'].mean()

In [ ]:
df[df['Embarked'] == 'S'].mean()

A: Highest average fares were paid by passengers originating from the Cherbourg port.

**What is the most common passenger class for each port?**

In [ ]:
cherbourg = df[df['Embarked'] == 'C']
cherbourg['Pclass'].value_counts()

In [ ]:
queenstown = df[df['Embarked'] == 'Q']
queenstown['Pclass'].value_counts()

In [ ]:
southampton = df[df['Embarked'] == 'S']
southampton['Pclass'].value_counts()

A: Cherbourg had mostly Upper class passengers, Queenstown and Southampton had mostly Lower class passengers.

**What fraction of surviving 1st-class males paid lower than double the overall median ticket price?**

In [ ]:
male_surv_1st = male_surv[male_surv['Pclass'] == 1]
len(male_surv_1st[male_surv_1st['Fare'] < (df['Fare'].median() * 2)])/len(male_surv_1st)

**How much older/younger was the average surviving passenger with family members than the average non-surviving passenger without them?**

In [ ]:
nonsurv_nofam = nonsurvivors[nonsurvivors['SibSp'] == 0]
survivors_wfam = survivors[survivors['SibSp'] != 0]
print('%f years younger' % (survivors_wfam['Age'].mean() - nonsurv_nofam['Age'].mean()))

**Display the relationship (i.e. make a plot) between survival rate and the quantile of the ticket price for 20 integer quantiles.**
- To be clearer, what I want is for you to specify 20 quantiles, and for each of those quantiles divide the number of survivors in that quantile by the total number of people in that quantile. That’ll give you the survival rate in that quantile.
- Then plot a line of the survival rate against the ticket fare quantiles.
- Make sure you label your axes.

In [ ]:
rate_list = []
j = 0

for i in range(20):
    q = len(survivors[survivors['Fare'] <= survivors['Fare'].quantile(j)])
    p = len(df[df['Fare'] <= df['Fare'].quantile(j)])
    diff = p - q
    rate = (diff/p if p!=0 else 0.0) 
    rate_list.append(rate)
    j = (i+1) * 0.05

plt.plot(rate_list)
plt.ylabel('Survival rate')
plt.xlabel('Fare')
plt.show()

**For each of the following characteristics, find the median in the data:**
- Age
- Ticket price
- Siblings/spouses
- Parents/children

In [ ]:
med_age = df['Age'].median()
med_fare = df['Fare'].median()
med_sibSp = df['SibSp'].median()
med_parch = df['Parch'].median()

**If you were to use these medians to draw numerical boundaries separating survivors from non-survivors, which of these characteristics would be the best choice and why?**

In [ ]:
# ** Not done yet **
surv_ages = survivors['Age'].dropna(axis=0, how='all')
nonsurv_ages = nonsurvivors['Age'].dropna(axis=0, how='all')

**Plot the distribution of passenger ages. Choose visually-meaningful bin sizes and label your axes.**

In [ ]:
all_ages = df['Age'].dropna(axis=0, how='all')
all_ages.plot(kind='hist', bins=8, figsize=(10, 10))
plt.vlines(med_age, ymin=0, ymax=230, linewidth=1.0, color="y")
plt.xlabel('Passenger Age')
plt.ylabel('Count')
plt.title('Reported Ages - All passengers')
plt.show()

**Find the probability that:**
- A passenger survived
- A passenger was male
- A passenger was female and had at least one sibling or spouse on board
- A survivor was from Cherbourg
- A passenger was less than 10 years old
- A passenger was between 25 and 40 years old
- A passenger was either younger than 20 years old or older than 50

In [ ]:
print('The probability that:')
print('A passenger survived: %f' % (len(survivors)/len(df)))
print('A passenger was male: %f' % (len(df[df['Sex'] == 'male'])/len(df)))
females = df[df['Sex'] == 'female']
print('A passenger was female and had at least one sibling or spouse on board: %f' % \
        (len(females[females['SibSp'] > 0])/len(df)))
print('A passenger from Cherbourg: %f' % (len(survivors[survivors['Embarked'] == 'C'])/len(df)))
print('A passenger was less than 10 years old: %f' % (len(df[df['Age'] < 10])/len(df)))
over_25 = df[df['Age'] > 25]
print('A passenger was between 25 and 40 years old: %f' % (len(over_25[over_25['Age'] < 40])/len(df)))
print('A passenger was either younger than 20 years old or older than 50: %f' % \
      ((len(df[df['Age'] < 20]) + len(df[df['Age'] > 50]))/len(df)))

**Knowing nothing else about the passengers aside from the survival rate of the population (see question above), if I choose 100 passengers at random from the passenger list, what’s the probability that exactly 42 passengers survive?**

In [ ]:
import scipy.stats  as stats

stats.binom.pmf(k=42, n=100, p=0.383838)

**What’s the probability that at least 42 of those 100 passengers survive?**

In [ ]:
1 - stats.binom.cdf(k=42, n=100, p=0.383838)

**Write a function that takes N random samples of 100 passengers, and returns the fraction of those samples where at least 42 passengers survive.**

Choose a random seed and find approximately how many random samples you need to take before your fraction matches the probability you calculated (within Δp≈0.05Δp≈0.05).

It may help to visualize the survival fraction vs the number of random samples. Answers will vary based on the seed.

In [ ]:
np.random.seed(10)
prob_estimates = []

def give_fraction(num):
    prob_estimates = []
    for x in range(num):
        sample = np.random.choice(a=df, size=100)
        point_estimates.append(sample.mean())

**Is there a statistically significant difference between:**
- The ages of male and female survivors?
- The fares paid by passengers from Queenstown and the passengers from Cherbourg?

**If so, at what level? If not, how do you know?**

**Accompany your p-values with histograms showing the distributions of both compared populations.**

**Did survivors pay more for their tickets than those that did not? If so, to what significance level?**

**Did a given first-class passenger have less family members on board than a given third-class passenger? If so, to what significance level?**